src/cli/self_update/windows.rs

```diff
@@ -633,47 +620,42 @@ pub(crate) fn do_add_to_programs(process: &Process) -> Result<()> {
     uninstall_cmd.push(path);
     uninstall_cmd.push("\" self uninstall");
 
-    let reg_value = RegValue {
-        bytes: to_winreg_bytes(uninstall_cmd.encode_wide().collect()),
-        vtype: RegType::REG_SZ,
-    };
-
-    key.set_raw_value("UninstallString", &reg_value)
-        .context("Failed to set `UninstallString`")?;
-    key.set_value("DisplayName", &"Rustup: the Rust toolchain installer")
+    key.set_bytes(
+        "UninstallString",
+        Type::String,
+        &to_winreg_bytes(uninstall_cmd.encode_wide().collect()),
+    )
+    .context("Failed to set `UninstallString`")?;
+    key.set_string("DisplayName", "Rustup: the Rust toolchain installer")
         .context("Failed to set `DisplayName`")?;
     do_update_programs_display_version(env!("CARGO_PKG_VERSION"))?;
 
     Ok(())
 }
 
 pub(crate) fn do_remove_from_programs() -> Result<()> {
-    match RegKey::predef(HKEY_CURRENT_USER).delete_subkey_all(RUSTUP_UNINSTALL_ENTRY) {
+    match CURRENT_USER.remove_tree(RUSTUP_UNINSTALL_ENTRY) {
         Ok(()) => Ok(()),
-        Err(ref e) if e.kind() == std::io::ErrorKind::NotFound => Ok(()),
+        Err(e) if e.code() == HRESULT::from_win32(ERROR_FILE_NOT_FOUND) => Ok(()),
         Err(e) => Err(anyhow!(e)),
     }
 }
 
 /// Convert a vector UCS-2 chars to a null-terminated UCS-2 string in bytes
 pub(crate) fn to_winreg_bytes(mut v: Vec<u16>) -> Vec<u8> {
     v.push(0);
-    unsafe { std::slice::from_raw_parts(v.as_ptr().cast::<u8>(), v.len() * 2).to_vec() }
+    unsafe { slice::from_raw_parts(v.as_ptr().cast::<u8>(), v.len() * 2).to_vec() }
 }
 
 /// This is used to decode the value of HKCU\Environment\PATH. If that key is
 /// not REG_SZ | REG_EXPAND_SZ then this returns None. The winreg library itself
 /// does a lossy unicode conversion.
-pub(crate) fn from_winreg_value(val: &winreg::RegValue) -> Option<Vec<u16>> {
-    use std::slice;
-
-    match val.vtype {
-        RegType::REG_SZ | RegType::REG_EXPAND_SZ => {
+pub(crate) fn from_winreg_value(val: Value) -> Option<Vec<u16>> {
```

Comment:
I wonder if `HSTRING` can also be used instead of `Vec<u16>` here to avoid `unsafe`. I'm not sure of this is feasible, just proposing a possibility... 🤔

Reply:
Maybe this can open a seperate PR?